### ISSUE TITLE
java.lang.VerifyError: Verifier rejected class com.clevertap.android.sdk.InAppFCManager

### ISSUE THREAD

**Issue by rishabh876** [CleverTap/clevertap-android-sdk#15]
Suddenly started getting this error on v3.3.3

> java.lang.VerifyError: Verifier rejected class com.clevertap.android.sdk.InAppFCManager: int[] com.clevertap.android.sdk.InAppFCManager.getInAppCountsFromPersistentStore(java.lang.String) failed to verify: int[] com.clevertap.android.sdk.InAppFCManager.getInAppCountsFromPersistentStore(java.lang.String): [0x3B] register v2 has type Conflict but expected Integer (declaration of 'com.clevertap.android.sdk.InAppFCManager' appears in /data/app/com.myapp.staging-w9oMEq_bIf4_tZ0J65nkLg==/base.apk)
>         at com.clevertap.android.sdk.CleverTapAPI.<init>(CleverTapAPI.java:182)
>         at com.clevertap.android.sdk.CleverTapAPI.instanceWithConfig(CleverTapAPI.java:456)
>         at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstance(CleverTapAPI.java:435)
>         at com.myapp.app.MyApp.initCleverTap(MyApp.kt:94)
>         at com.myapp.app.MyApp.onCreate(MyApp.kt:74)

This is what I am doing in my Application class `onCreate()`

```
private fun initCleverTap() {
        cleverTap = CleverTapAPI.getDefaultInstance(applicationContext)
        cleverTap?.enableDeviceNetworkInfoReporting(true)
        CleverTapAPI.setDebugLevel(CleverTapAPI.LogLevel.DEBUG)
    }
```

Android Studio 3.3 RC2

**Comment by rishabh876** [comment]
Update:
Working on Android Studio 3.2.1. Even though 3.3 is at release candidate 2.

**Comment by deeps2HOG** [comment]
just updated Android Studio to v3.3 and received this error.
```
com.tf.eros.faythTv E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.tf.eros.faythTv, PID: 31342
    java.lang.VerifyError: Verifier rejected class com.clevertap.android.sdk.InAppFCManager: int[] com.clevertap.android.sdk.InAppFCManager.getInAppCountsFromPersistentStore(java.lang.String) failed to verify: int[] com.clevertap.android.sdk.InAppFCManager.getInAppCountsFromPersistentStore(java.lang.String): [0x3B] register v2 has type Conflict but expected Integer (declaration of 'com.clevertap.android.sdk.InAppFCManager' appears in /data/app/com.tf.eros.faythTv-5h6epuM92Ex3XRWWF7NQZA==/split_lib_dependencies_apk.apk)
        at com.clevertap.android.sdk.CleverTapAPI.<init>(CleverTapAPI.java:225)
        at com.clevertap.android.sdk.CleverTapAPI.instanceWithConfig(CleverTapAPI.java:524)
        at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstance(CleverTapAPI.java:502)
        at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstanceOrFirstOther(CleverTapAPI.java:5008)
        at com.clevertap.android.sdk.CleverTapAPI.createNotificationChannel(CleverTapAPI.java:5544)
        at com.tf.eros.faythTv.application.AppApplication.setCleverTapNotificationChannel(AppApplication.java:498)
        at com.tf.eros.faythTv.application.AppApplication.onCreate(AppApplication.java:195)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1130)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5965)
        at android.app.ActivityThread.-wrap1(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1755)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6753)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:482)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

I am just creating a notification channel in my AppApplication File:
 ```
private void setCleverTapNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            CleverTapAPI.createNotificationChannel(getApplicationContext(),
                    AppConstants.CHANNEL_ID_MARKETING,
                    AppConstants.CHANNEL_NAME_MARKETING,
                    AppConstants.CHANNEL_DESCRIPTION_MARKETING,
                    NotificationManager.IMPORTANCE_MAX,
                    true);
    }
```
Any idea why crash is occurring ?

**Comment by calvarez-ov** [comment]
Getting this with:
* clevertap 3.3.4
* android gradle plugin 3.3.0
* gradle 5.1

I'm building on the comand line. I killed all gradle daemons and built with a clean:
```
./gradlew clean assembleDebug
```

**Comment by deeps2HOG** [comment]
> According to this [StackOverflow answer](https://stackoverflow.com/a/2518002/8301966), this seems to be like a project structure problem if it is occurring based on the version of Android Studio. Please `clean` and `rebuild project` once more or try `File -> Invalidate caches/ restart` and let us know if this problem still persists

tried both cleaning/rebuilding and invalidate caches/restart but still getting crash

**Comment by calvarez-ov** [comment]
Here's a sample app demonstrating the problem: https://github.com/libon/CleverTapVerifyError

Just build it and install it. It will crash.

**Comment by calvarez-ov** [comment]
An issue has been filed on the google issue tracker: https://issuetracker.google.com/issues/122887884

**Comment by darshanclevertap** [comment]
I was able to reproduce this issue on my end as well, so a big thank you to @calvarez-ov for raising the issue with Google.
According to the [issue tracker](https://issuetracker.google.com/issues/122887884) they have put in a fix but we will have to wait as to when it goes live.
Meanwhile as a workaround, downgrading the Gradle tools version to 3.2.1 will not cause this crash.
Make sure your Project's `build.gradle` file has the classpath dependency as follows -
`classpath 'com.android.tools.build:gradle:3.2.1'`
I will keep this thread open and keep an eye on the issue tracker for further updates.

**Comment by darshanclevertap** [comment]
As per the [issue tracker](https://issuetracker.google.com/issues/122887884), the fix for this issue is as follows - 
The fix for this issue is available for AGP(Android Gradle Plugin) 3.3 (and 3.4) by setting an explicit dependency detailed below.
After AGP 3.3.1 is released, remove the pinned version to allow you to pick up new D8/R8 releases again.

For AGP 3.3 amend your top-level `build.gradle` file with:

```
buildscript {
    repositories {
        maven {
            url "http://storage.googleapis.com/r8-releases/raw" // ADD THIS.
        }
    }
    dependencies {
        classpath 'com.android.tools:r8:1.3.52'  // ADD THIS. Must be before the Gradle Plugin for Android.
        classpath 'com.android.tools.build:gradle:3.3'
    }
}
```

For AGP 3.4 the the r8 version should be 1.4.25

I have tested this on my personal app integrated with CleverTap SDK and it works, hence closing this issue now. If you guys still face an issue let me know.

**Comment by rishabh876** [comment]
I also had to add maven `{ url "https://kotlin.bintray.com/kotlinx/" }` for 
`kotlinx-metadata-jvm:0.0.3` required by `r8:1.3.52`

**Comment by darshanclevertap** [comment]
Thank you @rishabh876  👍

**Comment by Jack2jm** [comment]
Same error getting this with:

clevertap 3.3.4
android gradle plugin 3.3.0
gradle (4.10.2).

java.lang.VerifyError: Verifier rejected class com.clevertap.android.sdk.InAppFCManager: int[] com.clevertap.android.sdk.InAppFCManager.getInAppCountsFromPersistentStore(java.lang.String) failed to verify: int[] com.clevertap.android.sdk.InAppFCManager.getInAppCountsFromPersistentStore(java.lang.String): [0x3B] register v2 has type Conflict but expected Integer (declaration of 'com.clevertap.android.sdk.InAppFCManager' appears in /data/app/com.amitech.allevents-mHvSMw6cEYHuy7e0ARcYQQ==/split_lib_dependencies_apk.apk)

**Comment by Jack2jm** [comment]
Same error getting this with:

clevertap 3.3.4
android gradle plugin 3.3.0
gradle (4.10.2).

java.lang.VerifyError: Verifier rejected class com.clevertap.android.sdk.InAppFCManager: int[] com.clevertap.android.sdk.InAppFCManager.getInAppCountsFromPersistentStore(java.lang.String) failed to verify: int[] com.clevertap.android.sdk.InAppFCManager.getInAppCountsFromPersistentStore(java.lang.String): [0x3B] register v2 has type Conflict but expected Integer (declaration of 'com.clevertap.android.sdk.InAppFCManager' appears in /data/app/package-mHvSMw6cEYHuy7e0ARcYQQ==/split_lib_dependencies_apk.apk)

Getting this with:

clevertap 3.3.4 (also occurs with 3.4.0)
android gradle plugin 3.3.0 or 3.4.0-alpha10 (no problem with 3.2.1)
gradle 5.1 (also occurs with 4.10.2)

**Comment by Jack2jm** [comment]
> As per the [issue tracker](https://issuetracker.google.com/issues/122887884), the fix for this issue is as follows -
> The fix for this issue is available for AGP(Android Gradle Plugin) 3.3 (and 3.4) by setting an explicit dependency detailed below.
> After AGP 3.3.1 is released, remove the pinned version to allow you to pick up new D8/R8 releases again.
> 
> For AGP 3.3 amend your top-level `build.gradle` file with:
> 
> ```
> buildscript {
>     repositories {
>         maven {
>             url "http://storage.googleapis.com/r8-releases/raw" // ADD THIS.
>         }
>     }
>     dependencies {
>         classpath 'com.android.tools:r8:1.3.52'  // ADD THIS. Must be before the Gradle Plugin for Android.
>         classpath 'com.android.tools.build:gradle:3.3'
>     }
> }
> ```
> For AGP 3.4 the the r8 version should be 1.4.25
> 
> I have tested this on my personal app integrated with CleverTap SDK and it works, hence closing this issue now. If you guys still face an issue let me know.

Tried this but it gives me error about kotlin dependencies, but i do not uses kotlin in my project.

**Comment by calvarez-ov** [comment]
@Jack2jm try adding it this way:
```
        classpath ("com.android.tools:r8:1.3.52" ) { transitive false }
```

**Comment by rishabh876** [comment]
Android Studio 3.3.1 is now available. So AGP 3.3.1 is also available. No need to use this work around anymore.

**Comment by darshanclevertap** [comment]
Thanks @rishabh876 👍 
I can confirm now that this issue doesn't occur if you update your AGP to `3.3.1`. Use this in your Project's `build.gradle`  dependencies -

`classpath 'com.android.tools.build:gradle:3.3.1'`

Thank you all for your contributions in helping solve this issue! 😃 👍

**Comment by shahsurajk** [comment]
If somehow had put this: 
`classpath ("com.android.tools:r8:1.3.52" ) { transitive false }` and updated to AS 3.4 and go a weird build error, 
> android.R8.getVersionString() not found etc

remove this dependency and sync.

**Comment by mkansara** [comment]
> As per the [issue tracker](https://issuetracker.google.com/issues/122887884), the fix for this issue is as follows -
> The fix for this issue is available for AGP(Android Gradle Plugin) 3.3 (and 3.4) by setting an explicit dependency detailed below.
> After AGP 3.3.1 is released, remove the pinned version to allow you to pick up new D8/R8 releases again.
> 
> For AGP 3.3 amend your top-level `build.gradle` file with:
> 
> ```
> buildscript {
>     repositories {
>         maven {
>             url "http://storage.googleapis.com/r8-releases/raw" // ADD THIS.
>         }
>     }
>     dependencies {
>         classpath 'com.android.tools:r8:1.3.52'  // ADD THIS. Must be before the Gradle Plugin for Android.
>         classpath 'com.android.tools.build:gradle:3.3'
>     }
> }
> ```
> 
> For AGP 3.4 the the r8 version should be 1.4.25
> 
> I have tested this on my personal app integrated with CleverTap SDK and it works, hence closing this issue now. If you guys still face an issue let me know.

Solve the issue with above solution

**Comment by Pragramer** [comment]
> As per the [issue tracker](https://issuetracker.google.com/issues/122887884), the fix for this issue is as follows -
> The fix for this issue is available for AGP(Android Gradle Plugin) 3.3 (and 3.4) by setting an explicit dependency detailed below.
> After AGP 3.3.1 is released, remove the pinned version to allow you to pick up new D8/R8 releases again.
> 
> For AGP 3.3 amend your top-level `build.gradle` file with:
> 
> ```
> buildscript {
>     repositories {
>         maven {
>             url "http://storage.googleapis.com/r8-releases/raw" // ADD THIS.
>         }
>     }
>     dependencies {
>         classpath 'com.android.tools:r8:1.3.52'  // ADD THIS. Must be before the Gradle Plugin for Android.
>         classpath 'com.android.tools.build:gradle:3.3'
>     }
> }
> ```
> 
> For AGP 3.4 the the r8 version should be 1.4.25
> 
> I have tested this on my personal app integrated with CleverTap SDK and it works, hence closing this issue now. If you guys still face an issue let me know.

i tried but useless..
classpath 'com.android.tools:r8:2.1.75'
classpath 'com.android.tools.build:gradle:3.6.3'

**Comment by darshanclevertap** [comment]
Please upgrade your AGP and move to the latest version of CleverTap Android SDK v4.0.2
